doc/HLD_of_SNS_Repair.rst

```diff
@@ -0,0 +1,1647 @@
+
+-  Number of storage devices attached to a server
+
+-  Number of servers in the pool
+
+-  Storage device bandwidth
+
+-  Storage device capacity
+
+-  Storage device space utilization
+
+-  Server-to-server network bandwidth
+
+-  Processor bandwidth per server
+
+-  Frequency and statistical distribution of client IO with the pool
+
+-  Frequency and distribution of permanent device failures
+
+-  Frequency and distribution of permanent server failures
+
+-  Frequency and distribution of transient server failures (restarts)
+
+-  Mean time to replace a storage device
+
+-  Mean time to replace a server
+
+-  Fraction of storage bandwidth used by repair
+
+-  Fraction of storage bandwidth used by rebalancing
+
+-  Fraction of network bandwidth used by repair
+
+-  Fraction of network bandwidth used by rebalancing
+
+-  Degradation in visible client IO rates
+
+-  Pool striping pattern: (N+K)/G
+
+Major SNS repair behavior metrics, affected by the above parameters are:
+
+-  Mean time to repair a storage device
+
+-  Mean time to repair a server
+
+-  Mean time to re-balance to a new storage device
+
+-  Mean time to re-balance to a new server
+
+To keep this section reasonably short, a number of simplifying assumptions, some of which can be easily lifted, are made:
+
+-  A pool consists of N\ :sub:`D` devices attached to N\ :sub:`S` servers (the same number of devices on each server)
+
+-  Every cluster-wide object is N+K striped across all servers and devices in the pool using parity declustering
+
+-  Device size is S\ :sub:`D` (bytes)
+
+-  Average device utilization (a fraction of used device space) is U
+
+-  Device bandwidth is B\ :sub:`D` (bytes/sec)
+
+-  Server-to-server network bandwidth is B\ :sub:`S` (bytes/sec)
+
+-  Server processor bandwidth is B\ :sub:`P` (defined as a rate at which RAID6 redundancy codes can be calculated, bytes/sec)
+
+-  Fractions of respectively storage, network and processor bandwidth dedicated to repair are A\ :sub:`S`, A\ :sub:`N` and A\ :sub:`P`
+
+Let's consider a steady state of repair in a pool with F\ :sub:`S` failed servers and F\ :sub:`D` failed devices (F\ :sub:`D` includes all
+devices on failed servers), assuming at most one unit is lost in any parity group. Define G\ :sub:`D`, G\ :sub:`S`, G\ :sub:`P` and G\ :sub:`O` as rates (bytes/sec) at which repair reads data from a device, sends data from a given server, computes redundancy codes and writes data to a device respectively. 
+
+Every one of N\ :sub:`S` - F\ :sub:`S` survived servers has on average (N\ :sub:`D` - F\ :sub:`D`)/(N\ :sub:`S` - F\ :sub:`S`) devices attached
+and from each of these data are read at the rate G\ :sub:`D`. Assuming that none of this data are for "internal consumption" (that is, assuming
+that no parity group has a spare space unit on a server where it has data or parity units),  servers sends out all these data, giving
+
+.. math:: G_{D}^{} \cdot \frac{N_{D}^{} - F_{D}^{}}{N_{S}^{} - F_{S}^{}} = G_{S}^{}
```

Comment:
Does this display properly in your web browser?

It does not show any meaningful math in my chrome.
Can you check?

Reply:
updated